python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py

```diff
@@ -0,0 +1,183 @@
+    start_list_dtype,
+    edge_dir,
+    fanout,
+    with_replacement,
+):
+    start_list_types = list(start_list_d.keys())
+    output_dfs = []
+    for can_etype, (sg, start_list_range) in sgs.items():
+        can_etype = _convert_can_etype_s_to_tup(can_etype)
+        if _edge_types_contains_canonical_etype(can_etype, start_list_types, edge_dir):
+            if edge_dir == "in":
+                subset_type = can_etype[2]
+            else:
+                subset_type = can_etype[0]
+            output = sample_single_sg(
+                sg,
+                sample_f,
+                start_list_d[subset_type],
+                start_list_dtype,
+                start_list_range,
+                fanout,
+                with_replacement,
+            )
+            output_dfs.append(output)
+
+    if len(output_dfs) == 0:
+        empty_df = cudf.DataFrame({"sources": [], "destinations": [], "indices": []})
+        return empty_df.astype(cp.int32)
+
+    if isinstance(output_dfs[0], dask_cudf.DataFrame):
+        return dask_cudf.concat(output_dfs, ignore_index=True)
+    else:
+        return cudf.concat(output_dfs, ignore_index=True)
+
+
+def sample_single_sg(
+    sg,
+    sample_f,
+    start_list,
+    start_list_dtype,
+    start_list_range,
+    fanout,
+    with_replacement,
+):
+    if isinstance(start_list, dict):
+        start_list = cudf.concat(list(start_list.values()))
+
+    # Uniform sampling fails when the dtype
+    # of the seed dtype is not same as the node dtype
+    start_list = start_list.astype(start_list_dtype)
+    # Filter start list by ranges
+    # https://github.com/rapidsai/cugraph/blob/branch-22.12/cpp/src/prims/per_v_random_select_transform_outgoing_e.cuh
+
+    start_list = start_list[
+        (start_list >= start_list_range[0]) & (start_list <= start_list_range[1])
+    ]
+
+    sampled_df = sample_f(
+        sg,
+        start_list=start_list,
+        fanout_vals=[fanout],
+        with_replacement=with_replacement,
+        # FIXME: is_edge_ids=True does not seem to do anything
+        # issue https://github.com/rapidsai/cugraph/issues/2562
+    )
+    return sampled_df
+
+
+def _edge_types_contains_canonical_etype(can_etype, edge_types, edge_dir):
+    src_type, _, dst_type = can_etype
+    if edge_dir == "in":
+        return dst_type in edge_types
+    else:
+        return src_type in edge_types
+
+
+def _convert_can_etype_s_to_tup(canonical_etype_s):
+    src_type, etype, dst_type = canonical_etype_s.split(",")
+    src_type = src_type[2:-1]
+    dst_type = dst_type[2:-2]
+    etype = etype[2:-1]
+    return (src_type, etype, dst_type)
+
+
+def create_dlpack_d(d):
+    dlpack_d = {}
+    for k, df in d.items():
+        if len(df) == 0:
+            dlpack_d[k] = (None, None, None)
+        else:
+            dlpack_d[k] = (
+                df[src_n].to_dlpack(),
+                df[dst_n].to_dlpack(),
+                df[eid_n].to_dlpack(),
+            )
+
+    return dlpack_d
```

Comment:
Moved from cugraph_store and renamed to `create_dlpack_d`. 
 https://github.com/rapidsai/cugraph/blob/fbd5f204989364af9d30ab2b4d8ad72e276ced0a/python/cugraph/cugraph/gnn/graph_store.py#L586-L598